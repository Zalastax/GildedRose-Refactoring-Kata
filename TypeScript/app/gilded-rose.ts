export class Item {
  name: string;
  sellIn: number;
  quality: number;

  constructor(name, sellIn, quality) {
    this.name = name;
    this.sellIn = sellIn;
    this.quality = quality;
  }
}

function defaultItemQualityChange(sellIn: number) {
  if (sellIn < 0) {
    return -2;
  }
  return -1;
}

function updateDefaultItem(item: Item) {
  updateItem(defaultItemQualityChange, item)
}

function doubleItemQualityChange(sellIn: number) {
  return defaultItemQualityChange(sellIn) * 2;
}

function updateConjured(item: Item) {
  updateItem(doubleItemQualityChange, item)
}

function backstageQualityChange(sellIn: number) {
  if (sellIn < 6) {
    return 3;
  }
  if (sellIn < 11) {
    return 2;
  }
  return 1;
}

function updateBackstageItem(item: Item) {
  // It's a bit strange that this item modifies sellIn
  // after updating quality while others update it before...
  item.quality = Math.min(
    50,
    item.quality + backstageQualityChange(item.sellIn)
  );

  item.sellIn = item.sellIn - 1;
  if (item.sellIn < 0) {
    item.quality = 0;
  }
}

function brieItemQualityChange(sellIn: number) {
  if (sellIn < 0) {
    return 2;
  }
  return 1;
}

function updateBrieItem(item: Item) {
  updateItem(brieItemQualityChange, item)
}

type QualityChange = (sellIn: number) => number;

function updateItem(qualityChange: QualityChange, item: Item) {
    item.sellIn = item.sellIn - 1;
    item.quality = Math.max(0, Math.min(50, item.quality + qualityChange(item.sellIn)));
}

function itemNoOp(item: Item) {}

type SideEffect = () => void;

const updatersMapping = new Map([
  ["Aged Brie", updateBrieItem],
  ["Backstage passes to a TAFKAL80ETC concert", updateBackstageItem],
  ["Sulfuras, Hand of Ragnaros", itemNoOp],
  // Looking at other implementations it seems like the only conjured item is
  // Conjured Mana Cake. Would be easy to change from map to a function if more conjured items
  // should be supported.
  ["Conjured Mana Cake", updateConjured]
]);

function getOrDefault<K, V>(map: Map<K, V>, key: K, orDefault: V) {
  if (map.has(key)) {
    return map.get(key);
  }
  return orDefault;
}

export class GildedRose {
  private items: Array<Item>;
  private itemUpdaters: Array<SideEffect>;

  /**
   *
   * @param items Items to update.
   * Passing items to this constructor gives up ownership;
   * do not modify items outside GildedRose!
   */
  constructor(items: Array<Item>) {
    this.items = items;
    this.itemUpdaters = items.map(item => {
      // Select update function via its name
      const updater = getOrDefault(
        updatersMapping,
        item.name,
        updateDefaultItem
      );
      // Bind the item parameter to the update function
      return updater.bind(undefined, item);
    });
  }

  updateQuality() {
    for (const updater of this.itemUpdaters) {
      updater();
    }

    return this.items;
  }
}
