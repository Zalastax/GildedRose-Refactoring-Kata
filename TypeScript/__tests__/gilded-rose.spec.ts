import { Item, GildedRose } from "../app/gilded-rose";

describe("Gilded Rose", function() {
  const itemNames = [
    "foo",
    "Aged Brie",
    "Backstage passes to a TAFKAL80ETC concert",
    "Sulfuras, Hand of Ragnaros",
    "Conjured Mana Cake"
  ];
  const sellIns = [0, 1, -1, 6, 11];
  const qualities = [0, 1, 50, 49];

  for (const name of itemNames) {
    for (const sellIn of sellIns) {
      for (const quality of qualities) {
        it(`${name} of quality ${quality} sellsIn ${sellIn}`, function() {
          const gildedRose = new GildedRose([new Item(name, sellIn, quality)]);
          const items = gildedRose.updateQuality();
          expect(items[0]).toMatchSnapshot();
        });
      }
    }
  }
});
