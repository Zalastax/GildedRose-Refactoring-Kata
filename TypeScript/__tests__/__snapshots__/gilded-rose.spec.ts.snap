// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gilded Rose Aged Brie of quality 0 sellsIn -1 1`] = `
Item {
  "name": "Aged Brie",
  "quality": 2,
  "sellIn": -2,
}
`;

exports[`Gilded Rose Aged Brie of quality 0 sellsIn 0 1`] = `
Item {
  "name": "Aged Brie",
  "quality": 2,
  "sellIn": -1,
}
`;

exports[`Gilded Rose Aged Brie of quality 0 sellsIn 1 1`] = `
Item {
  "name": "Aged Brie",
  "quality": 1,
  "sellIn": 0,
}
`;

exports[`Gilded Rose Aged Brie of quality 0 sellsIn 6 1`] = `
Item {
  "name": "Aged Brie",
  "quality": 1,
  "sellIn": 5,
}
`;

exports[`Gilded Rose Aged Brie of quality 0 sellsIn 11 1`] = `
Item {
  "name": "Aged Brie",
  "quality": 1,
  "sellIn": 10,
}
`;

exports[`Gilded Rose Aged Brie of quality 1 sellsIn -1 1`] = `
Item {
  "name": "Aged Brie",
  "quality": 3,
  "sellIn": -2,
}
`;

exports[`Gilded Rose Aged Brie of quality 1 sellsIn 0 1`] = `
Item {
  "name": "Aged Brie",
  "quality": 3,
  "sellIn": -1,
}
`;

exports[`Gilded Rose Aged Brie of quality 1 sellsIn 1 1`] = `
Item {
  "name": "Aged Brie",
  "quality": 2,
  "sellIn": 0,
}
`;

exports[`Gilded Rose Aged Brie of quality 1 sellsIn 6 1`] = `
Item {
  "name": "Aged Brie",
  "quality": 2,
  "sellIn": 5,
}
`;

exports[`Gilded Rose Aged Brie of quality 1 sellsIn 11 1`] = `
Item {
  "name": "Aged Brie",
  "quality": 2,
  "sellIn": 10,
}
`;

exports[`Gilded Rose Aged Brie of quality 49 sellsIn -1 1`] = `
Item {
  "name": "Aged Brie",
  "quality": 50,
  "sellIn": -2,
}
`;

exports[`Gilded Rose Aged Brie of quality 49 sellsIn 0 1`] = `
Item {
  "name": "Aged Brie",
  "quality": 50,
  "sellIn": -1,
}
`;

exports[`Gilded Rose Aged Brie of quality 49 sellsIn 1 1`] = `
Item {
  "name": "Aged Brie",
  "quality": 50,
  "sellIn": 0,
}
`;

exports[`Gilded Rose Aged Brie of quality 49 sellsIn 6 1`] = `
Item {
  "name": "Aged Brie",
  "quality": 50,
  "sellIn": 5,
}
`;

exports[`Gilded Rose Aged Brie of quality 49 sellsIn 11 1`] = `
Item {
  "name": "Aged Brie",
  "quality": 50,
  "sellIn": 10,
}
`;

exports[`Gilded Rose Aged Brie of quality 50 sellsIn -1 1`] = `
Item {
  "name": "Aged Brie",
  "quality": 50,
  "sellIn": -2,
}
`;

exports[`Gilded Rose Aged Brie of quality 50 sellsIn 0 1`] = `
Item {
  "name": "Aged Brie",
  "quality": 50,
  "sellIn": -1,
}
`;

exports[`Gilded Rose Aged Brie of quality 50 sellsIn 1 1`] = `
Item {
  "name": "Aged Brie",
  "quality": 50,
  "sellIn": 0,
}
`;

exports[`Gilded Rose Aged Brie of quality 50 sellsIn 6 1`] = `
Item {
  "name": "Aged Brie",
  "quality": 50,
  "sellIn": 5,
}
`;

exports[`Gilded Rose Aged Brie of quality 50 sellsIn 11 1`] = `
Item {
  "name": "Aged Brie",
  "quality": 50,
  "sellIn": 10,
}
`;

exports[`Gilded Rose Backstage passes to a TAFKAL80ETC concert of quality 0 sellsIn -1 1`] = `
Item {
  "name": "Backstage passes to a TAFKAL80ETC concert",
  "quality": 0,
  "sellIn": -2,
}
`;

exports[`Gilded Rose Backstage passes to a TAFKAL80ETC concert of quality 0 sellsIn 0 1`] = `
Item {
  "name": "Backstage passes to a TAFKAL80ETC concert",
  "quality": 0,
  "sellIn": -1,
}
`;

exports[`Gilded Rose Backstage passes to a TAFKAL80ETC concert of quality 0 sellsIn 1 1`] = `
Item {
  "name": "Backstage passes to a TAFKAL80ETC concert",
  "quality": 3,
  "sellIn": 0,
}
`;

exports[`Gilded Rose Backstage passes to a TAFKAL80ETC concert of quality 0 sellsIn 6 1`] = `
Item {
  "name": "Backstage passes to a TAFKAL80ETC concert",
  "quality": 2,
  "sellIn": 5,
}
`;

exports[`Gilded Rose Backstage passes to a TAFKAL80ETC concert of quality 0 sellsIn 11 1`] = `
Item {
  "name": "Backstage passes to a TAFKAL80ETC concert",
  "quality": 1,
  "sellIn": 10,
}
`;

exports[`Gilded Rose Backstage passes to a TAFKAL80ETC concert of quality 1 sellsIn -1 1`] = `
Item {
  "name": "Backstage passes to a TAFKAL80ETC concert",
  "quality": 0,
  "sellIn": -2,
}
`;

exports[`Gilded Rose Backstage passes to a TAFKAL80ETC concert of quality 1 sellsIn 0 1`] = `
Item {
  "name": "Backstage passes to a TAFKAL80ETC concert",
  "quality": 0,
  "sellIn": -1,
}
`;

exports[`Gilded Rose Backstage passes to a TAFKAL80ETC concert of quality 1 sellsIn 1 1`] = `
Item {
  "name": "Backstage passes to a TAFKAL80ETC concert",
  "quality": 4,
  "sellIn": 0,
}
`;

exports[`Gilded Rose Backstage passes to a TAFKAL80ETC concert of quality 1 sellsIn 6 1`] = `
Item {
  "name": "Backstage passes to a TAFKAL80ETC concert",
  "quality": 3,
  "sellIn": 5,
}
`;

exports[`Gilded Rose Backstage passes to a TAFKAL80ETC concert of quality 1 sellsIn 11 1`] = `
Item {
  "name": "Backstage passes to a TAFKAL80ETC concert",
  "quality": 2,
  "sellIn": 10,
}
`;

exports[`Gilded Rose Backstage passes to a TAFKAL80ETC concert of quality 49 sellsIn -1 1`] = `
Item {
  "name": "Backstage passes to a TAFKAL80ETC concert",
  "quality": 0,
  "sellIn": -2,
}
`;

exports[`Gilded Rose Backstage passes to a TAFKAL80ETC concert of quality 49 sellsIn 0 1`] = `
Item {
  "name": "Backstage passes to a TAFKAL80ETC concert",
  "quality": 0,
  "sellIn": -1,
}
`;

exports[`Gilded Rose Backstage passes to a TAFKAL80ETC concert of quality 49 sellsIn 1 1`] = `
Item {
  "name": "Backstage passes to a TAFKAL80ETC concert",
  "quality": 50,
  "sellIn": 0,
}
`;

exports[`Gilded Rose Backstage passes to a TAFKAL80ETC concert of quality 49 sellsIn 6 1`] = `
Item {
  "name": "Backstage passes to a TAFKAL80ETC concert",
  "quality": 50,
  "sellIn": 5,
}
`;

exports[`Gilded Rose Backstage passes to a TAFKAL80ETC concert of quality 49 sellsIn 11 1`] = `
Item {
  "name": "Backstage passes to a TAFKAL80ETC concert",
  "quality": 50,
  "sellIn": 10,
}
`;

exports[`Gilded Rose Backstage passes to a TAFKAL80ETC concert of quality 50 sellsIn -1 1`] = `
Item {
  "name": "Backstage passes to a TAFKAL80ETC concert",
  "quality": 0,
  "sellIn": -2,
}
`;

exports[`Gilded Rose Backstage passes to a TAFKAL80ETC concert of quality 50 sellsIn 0 1`] = `
Item {
  "name": "Backstage passes to a TAFKAL80ETC concert",
  "quality": 0,
  "sellIn": -1,
}
`;

exports[`Gilded Rose Backstage passes to a TAFKAL80ETC concert of quality 50 sellsIn 1 1`] = `
Item {
  "name": "Backstage passes to a TAFKAL80ETC concert",
  "quality": 50,
  "sellIn": 0,
}
`;

exports[`Gilded Rose Backstage passes to a TAFKAL80ETC concert of quality 50 sellsIn 6 1`] = `
Item {
  "name": "Backstage passes to a TAFKAL80ETC concert",
  "quality": 50,
  "sellIn": 5,
}
`;

exports[`Gilded Rose Backstage passes to a TAFKAL80ETC concert of quality 50 sellsIn 11 1`] = `
Item {
  "name": "Backstage passes to a TAFKAL80ETC concert",
  "quality": 50,
  "sellIn": 10,
}
`;

exports[`Gilded Rose Conjured Mana Cake of quality 0 sellsIn -1 1`] = `
Item {
  "name": "Conjured Mana Cake",
  "quality": 0,
  "sellIn": -2,
}
`;

exports[`Gilded Rose Conjured Mana Cake of quality 0 sellsIn 0 1`] = `
Item {
  "name": "Conjured Mana Cake",
  "quality": 0,
  "sellIn": -1,
}
`;

exports[`Gilded Rose Conjured Mana Cake of quality 0 sellsIn 1 1`] = `
Item {
  "name": "Conjured Mana Cake",
  "quality": 0,
  "sellIn": 0,
}
`;

exports[`Gilded Rose Conjured Mana Cake of quality 0 sellsIn 6 1`] = `
Item {
  "name": "Conjured Mana Cake",
  "quality": 0,
  "sellIn": 5,
}
`;

exports[`Gilded Rose Conjured Mana Cake of quality 0 sellsIn 11 1`] = `
Item {
  "name": "Conjured Mana Cake",
  "quality": 0,
  "sellIn": 10,
}
`;

exports[`Gilded Rose Conjured Mana Cake of quality 1 sellsIn -1 1`] = `
Item {
  "name": "Conjured Mana Cake",
  "quality": 0,
  "sellIn": -2,
}
`;

exports[`Gilded Rose Conjured Mana Cake of quality 1 sellsIn 0 1`] = `
Item {
  "name": "Conjured Mana Cake",
  "quality": 0,
  "sellIn": -1,
}
`;

exports[`Gilded Rose Conjured Mana Cake of quality 1 sellsIn 1 1`] = `
Item {
  "name": "Conjured Mana Cake",
  "quality": 0,
  "sellIn": 0,
}
`;

exports[`Gilded Rose Conjured Mana Cake of quality 1 sellsIn 6 1`] = `
Item {
  "name": "Conjured Mana Cake",
  "quality": 0,
  "sellIn": 5,
}
`;

exports[`Gilded Rose Conjured Mana Cake of quality 1 sellsIn 11 1`] = `
Item {
  "name": "Conjured Mana Cake",
  "quality": 0,
  "sellIn": 10,
}
`;

exports[`Gilded Rose Conjured Mana Cake of quality 49 sellsIn -1 1`] = `
Item {
  "name": "Conjured Mana Cake",
  "quality": 45,
  "sellIn": -2,
}
`;

exports[`Gilded Rose Conjured Mana Cake of quality 49 sellsIn 0 1`] = `
Item {
  "name": "Conjured Mana Cake",
  "quality": 45,
  "sellIn": -1,
}
`;

exports[`Gilded Rose Conjured Mana Cake of quality 49 sellsIn 1 1`] = `
Item {
  "name": "Conjured Mana Cake",
  "quality": 47,
  "sellIn": 0,
}
`;

exports[`Gilded Rose Conjured Mana Cake of quality 49 sellsIn 6 1`] = `
Item {
  "name": "Conjured Mana Cake",
  "quality": 47,
  "sellIn": 5,
}
`;

exports[`Gilded Rose Conjured Mana Cake of quality 49 sellsIn 11 1`] = `
Item {
  "name": "Conjured Mana Cake",
  "quality": 47,
  "sellIn": 10,
}
`;

exports[`Gilded Rose Conjured Mana Cake of quality 50 sellsIn -1 1`] = `
Item {
  "name": "Conjured Mana Cake",
  "quality": 46,
  "sellIn": -2,
}
`;

exports[`Gilded Rose Conjured Mana Cake of quality 50 sellsIn 0 1`] = `
Item {
  "name": "Conjured Mana Cake",
  "quality": 46,
  "sellIn": -1,
}
`;

exports[`Gilded Rose Conjured Mana Cake of quality 50 sellsIn 1 1`] = `
Item {
  "name": "Conjured Mana Cake",
  "quality": 48,
  "sellIn": 0,
}
`;

exports[`Gilded Rose Conjured Mana Cake of quality 50 sellsIn 6 1`] = `
Item {
  "name": "Conjured Mana Cake",
  "quality": 48,
  "sellIn": 5,
}
`;

exports[`Gilded Rose Conjured Mana Cake of quality 50 sellsIn 11 1`] = `
Item {
  "name": "Conjured Mana Cake",
  "quality": 48,
  "sellIn": 10,
}
`;

exports[`Gilded Rose Sulfuras, Hand of Ragnaros of quality 0 sellsIn -1 1`] = `
Item {
  "name": "Sulfuras, Hand of Ragnaros",
  "quality": 0,
  "sellIn": -1,
}
`;

exports[`Gilded Rose Sulfuras, Hand of Ragnaros of quality 0 sellsIn 0 1`] = `
Item {
  "name": "Sulfuras, Hand of Ragnaros",
  "quality": 0,
  "sellIn": 0,
}
`;

exports[`Gilded Rose Sulfuras, Hand of Ragnaros of quality 0 sellsIn 1 1`] = `
Item {
  "name": "Sulfuras, Hand of Ragnaros",
  "quality": 0,
  "sellIn": 1,
}
`;

exports[`Gilded Rose Sulfuras, Hand of Ragnaros of quality 0 sellsIn 6 1`] = `
Item {
  "name": "Sulfuras, Hand of Ragnaros",
  "quality": 0,
  "sellIn": 6,
}
`;

exports[`Gilded Rose Sulfuras, Hand of Ragnaros of quality 0 sellsIn 11 1`] = `
Item {
  "name": "Sulfuras, Hand of Ragnaros",
  "quality": 0,
  "sellIn": 11,
}
`;

exports[`Gilded Rose Sulfuras, Hand of Ragnaros of quality 1 sellsIn -1 1`] = `
Item {
  "name": "Sulfuras, Hand of Ragnaros",
  "quality": 1,
  "sellIn": -1,
}
`;

exports[`Gilded Rose Sulfuras, Hand of Ragnaros of quality 1 sellsIn 0 1`] = `
Item {
  "name": "Sulfuras, Hand of Ragnaros",
  "quality": 1,
  "sellIn": 0,
}
`;

exports[`Gilded Rose Sulfuras, Hand of Ragnaros of quality 1 sellsIn 1 1`] = `
Item {
  "name": "Sulfuras, Hand of Ragnaros",
  "quality": 1,
  "sellIn": 1,
}
`;

exports[`Gilded Rose Sulfuras, Hand of Ragnaros of quality 1 sellsIn 6 1`] = `
Item {
  "name": "Sulfuras, Hand of Ragnaros",
  "quality": 1,
  "sellIn": 6,
}
`;

exports[`Gilded Rose Sulfuras, Hand of Ragnaros of quality 1 sellsIn 11 1`] = `
Item {
  "name": "Sulfuras, Hand of Ragnaros",
  "quality": 1,
  "sellIn": 11,
}
`;

exports[`Gilded Rose Sulfuras, Hand of Ragnaros of quality 49 sellsIn -1 1`] = `
Item {
  "name": "Sulfuras, Hand of Ragnaros",
  "quality": 49,
  "sellIn": -1,
}
`;

exports[`Gilded Rose Sulfuras, Hand of Ragnaros of quality 49 sellsIn 0 1`] = `
Item {
  "name": "Sulfuras, Hand of Ragnaros",
  "quality": 49,
  "sellIn": 0,
}
`;

exports[`Gilded Rose Sulfuras, Hand of Ragnaros of quality 49 sellsIn 1 1`] = `
Item {
  "name": "Sulfuras, Hand of Ragnaros",
  "quality": 49,
  "sellIn": 1,
}
`;

exports[`Gilded Rose Sulfuras, Hand of Ragnaros of quality 49 sellsIn 6 1`] = `
Item {
  "name": "Sulfuras, Hand of Ragnaros",
  "quality": 49,
  "sellIn": 6,
}
`;

exports[`Gilded Rose Sulfuras, Hand of Ragnaros of quality 49 sellsIn 11 1`] = `
Item {
  "name": "Sulfuras, Hand of Ragnaros",
  "quality": 49,
  "sellIn": 11,
}
`;

exports[`Gilded Rose Sulfuras, Hand of Ragnaros of quality 50 sellsIn -1 1`] = `
Item {
  "name": "Sulfuras, Hand of Ragnaros",
  "quality": 50,
  "sellIn": -1,
}
`;

exports[`Gilded Rose Sulfuras, Hand of Ragnaros of quality 50 sellsIn 0 1`] = `
Item {
  "name": "Sulfuras, Hand of Ragnaros",
  "quality": 50,
  "sellIn": 0,
}
`;

exports[`Gilded Rose Sulfuras, Hand of Ragnaros of quality 50 sellsIn 1 1`] = `
Item {
  "name": "Sulfuras, Hand of Ragnaros",
  "quality": 50,
  "sellIn": 1,
}
`;

exports[`Gilded Rose Sulfuras, Hand of Ragnaros of quality 50 sellsIn 6 1`] = `
Item {
  "name": "Sulfuras, Hand of Ragnaros",
  "quality": 50,
  "sellIn": 6,
}
`;

exports[`Gilded Rose Sulfuras, Hand of Ragnaros of quality 50 sellsIn 11 1`] = `
Item {
  "name": "Sulfuras, Hand of Ragnaros",
  "quality": 50,
  "sellIn": 11,
}
`;

exports[`Gilded Rose foo of quality 0 sellsIn -1 1`] = `
Item {
  "name": "foo",
  "quality": 0,
  "sellIn": -2,
}
`;

exports[`Gilded Rose foo of quality 0 sellsIn 0 1`] = `
Item {
  "name": "foo",
  "quality": 0,
  "sellIn": -1,
}
`;

exports[`Gilded Rose foo of quality 0 sellsIn 1 1`] = `
Item {
  "name": "foo",
  "quality": 0,
  "sellIn": 0,
}
`;

exports[`Gilded Rose foo of quality 0 sellsIn 6 1`] = `
Item {
  "name": "foo",
  "quality": 0,
  "sellIn": 5,
}
`;

exports[`Gilded Rose foo of quality 0 sellsIn 11 1`] = `
Item {
  "name": "foo",
  "quality": 0,
  "sellIn": 10,
}
`;

exports[`Gilded Rose foo of quality 1 sellsIn -1 1`] = `
Item {
  "name": "foo",
  "quality": 0,
  "sellIn": -2,
}
`;

exports[`Gilded Rose foo of quality 1 sellsIn 0 1`] = `
Item {
  "name": "foo",
  "quality": 0,
  "sellIn": -1,
}
`;

exports[`Gilded Rose foo of quality 1 sellsIn 1 1`] = `
Item {
  "name": "foo",
  "quality": 0,
  "sellIn": 0,
}
`;

exports[`Gilded Rose foo of quality 1 sellsIn 6 1`] = `
Item {
  "name": "foo",
  "quality": 0,
  "sellIn": 5,
}
`;

exports[`Gilded Rose foo of quality 1 sellsIn 11 1`] = `
Item {
  "name": "foo",
  "quality": 0,
  "sellIn": 10,
}
`;

exports[`Gilded Rose foo of quality 49 sellsIn -1 1`] = `
Item {
  "name": "foo",
  "quality": 47,
  "sellIn": -2,
}
`;

exports[`Gilded Rose foo of quality 49 sellsIn 0 1`] = `
Item {
  "name": "foo",
  "quality": 47,
  "sellIn": -1,
}
`;

exports[`Gilded Rose foo of quality 49 sellsIn 1 1`] = `
Item {
  "name": "foo",
  "quality": 48,
  "sellIn": 0,
}
`;

exports[`Gilded Rose foo of quality 49 sellsIn 6 1`] = `
Item {
  "name": "foo",
  "quality": 48,
  "sellIn": 5,
}
`;

exports[`Gilded Rose foo of quality 49 sellsIn 11 1`] = `
Item {
  "name": "foo",
  "quality": 48,
  "sellIn": 10,
}
`;

exports[`Gilded Rose foo of quality 50 sellsIn -1 1`] = `
Item {
  "name": "foo",
  "quality": 48,
  "sellIn": -2,
}
`;

exports[`Gilded Rose foo of quality 50 sellsIn 0 1`] = `
Item {
  "name": "foo",
  "quality": 48,
  "sellIn": -1,
}
`;

exports[`Gilded Rose foo of quality 50 sellsIn 1 1`] = `
Item {
  "name": "foo",
  "quality": 49,
  "sellIn": 0,
}
`;

exports[`Gilded Rose foo of quality 50 sellsIn 6 1`] = `
Item {
  "name": "foo",
  "quality": 49,
  "sellIn": 5,
}
`;

exports[`Gilded Rose foo of quality 50 sellsIn 11 1`] = `
Item {
  "name": "foo",
  "quality": 49,
  "sellIn": 10,
}
`;
